// feathers-mongoose/lib/service.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = init;

var _lodash = require('lodash.omit');

var _lodash2 = _interopRequireDefault(_lodash);

var _uberproto = require('uberproto');

var _uberproto2 = _interopRequireDefault(_uberproto);

var _feathersQueryFilters = require('feathers-query-filters');

var _feathersQueryFilters2 = _interopRequireDefault(_feathersQueryFilters);

var _feathersErrors = require('feathers-errors');

var _feathersErrors2 = _interopRequireDefault(_feathersErrors);

var _errorHandler = require('./error-handler');

var _errorHandler2 = _interopRequireDefault(_errorHandler);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// Create the service.

var Service = function () {
  function Service(options) {
    _classCallCheck(this, Service);

    if (!options) {
      throw new Error('Mongoose options have to be provided');
    }

    if (!options.Model || !options.Model.modelName) {
      throw new Error('You must provide a Mongoose Model');
    }

    // this.name = options.name;
    this.Model = options.Model;
    this.id = options.id || '_id';
    this.paginate = options.paginate || {};
    this.lean = options.lean || false;
    this.overwrite = options.overwrite === false ? false : true;
  }

  _createClass(Service, [{
    key: 'extend',
    value: function extend(obj) {
      return _uberproto2.default.extend(obj, this);
    }
  }, {
    key: '_find',
    value: function _find(params, count) {
      var getFilter = arguments.length <= 2 || arguments[2] === undefined ? _feathersQueryFilters2.default : arguments[2];

      var _getFilter = getFilter(params.query || {});

      var filters = _getFilter.filters;
      var query = _getFilter.query;

      var q = this.Model.find(query).lean(this.lean);

      // $select uses a specific find syntax, so it has to come first.
      if (filters.$select && filters.$select.length) {
        var fields = {};

        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = filters.$select[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var key = _step.value;

            fields[key] = 1;
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        q.select(fields);
      }

      // Handle $sort
      if (filters.$sort) {
        q.sort(filters.$sort);
      }

      // Handle $limit
      if (filters.$limit) {
        q.limit(filters.$limit);
      }

      // Handle $skip
      if (filters.$skip) {
        q.skip(filters.$skip);
      }

      // Handle $populate
      if (filters.$populate) {
        q.populate(filters.$populate);
      }

      var executeQuery = function executeQuery(total) {
        return q.exec().then(function (data) {
          return {
            total: total,
            limit: filters.$limit,
            skip: filters.$skip || 0,
            data: data
          };
        });
      };

      if (count) {
        return this.Model.where(query).count().exec().then(executeQuery);
      }

      return executeQuery();
    }
  }, {
    key: 'find',
    value: function find(params) {
      var paginate = params && typeof params.paginate !== 'undefined' ? params.paginate : this.paginate;
      var result = this._find(params, !!paginate.default, function (query) {
        return (0, _feathersQueryFilters2.default)(query, paginate);
      });

      if (!paginate.default) {
        return result.then(function (page) {
          return page.data;
        });
      }

      return result;
    }
  }, {
    key: '_get',
    value: function _get(id, params) {
      var modelQuery = this.Model.findOne(_defineProperty({}, this.id, id));

      if (params && params.query && params.query.$populate) {
        modelQuery = modelQuery.populate(params.query.$populate);
      }

      return modelQuery.lean(this.lean).exec().then(function (data) {
        if (!data) {
          throw new _feathersErrors2.default.NotFound('No record found for id \'' + id + '\'');
        }

        return data;
      }).catch(_errorHandler2.default);
    }
  }, {
    key: 'get',
    value: function get(id, params) {
      return this._get(id, params);
    }
  }, {
    key: '_getOrFind',
    value: function _getOrFind(id, params) {
      if (id === null) {
        return this._find(params).then(function (page) {
          return page.data;
        });
      }

      return this._get(id);
    }
  }, {
    key: 'create',
    value: function create(data) {
      return this.Model.create(data).catch(_errorHandler2.default);
    }
  }, {
    key: 'update',
    value: function update(id, data, params) {
      if (id === null) {
        return Promise.reject('Not replacing multiple records. Did you mean `patch`?');
      }

      var options = Object.assign({
        new: true,
        overwrite: this.overwrite,
        runValidators: true,
        context: 'query',
        setDefaultsOnInsert: true
      }, params.mongoose);

      if (this.id === '_id') {
        // We can not update default mongo ids
        data = (0, _lodash2.default)(data, this.id);
      } else {
        // If not using the default Mongo _id field set the id to its
        // previous value. This prevents orphaned documents.
        data = Object.assign({}, data, _defineProperty({}, this.id, id));
      }

      return this.Model.findOneAndUpdate(_defineProperty({}, this.id, id), data, options).lean(this.lean).exec().catch(_errorHandler2.default);
    }
  }, {
    key: 'patch',
    value: function patch(id, data, params) {
      var _this = this;

      params.query = params.query || {};
      data = Object.assign({}, data);

      // If we are updating multiple records
      var options = Object.assign({
        multi: id === null,
        runValidators: true,
        context: 'query'
      }, params.mongoose);

      if (id !== null) {
        params.query[this.id] = id;
      }

      if (this.id === '_id') {
        // We can not update default mongo ids
        delete data[this.id];
      } else {
        // If not using the default Mongo _id field set the id to its
        // previous value. This prevents orphaned documents.
        data[this.id] = id;
      }

      // We need this shitty hack because update doesn't return
      // a promise properly when runValidators is true. WTF!
      try {
        return this.Model.update(params.query, data, options).lean(this.lean).exec().then(function () {
          return _this._getOrFind(id, params);
        }).catch(_errorHandler2.default);
      } catch (e) {
        return (0, _errorHandler2.default)(e);
      }
    }
  }, {
    key: 'remove',
    value: function remove(id, params) {
      var _this2 = this;

      var query = Object.assign({}, params.query);

      if (id !== null) {
        query[this.id] = id;
      }

      // NOTE (EK): First fetch the record(s) so that we can return
      // it/them when we delete it/them.
      return this._getOrFind(id, params).then(function (data) {
        return _this2.Model.remove(query).lean(_this2.lean).exec().then(function () {
          return data;
        }).catch(_errorHandler2.default);
      }).catch(_errorHandler2.default);
    }
  }]);

  return Service;
}();

function init(options) {
  return new Service(options);
}

init.Service = Service;
module.exports = exports['default'];
